core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/RequesterPaysSetupServiceSpec.scala

```diff
@@ -0,0 +1,90 @@
+
+    service.getBondProviderServiceAccountEmails(userInfo).futureValue shouldBe List.empty
+  }
+
+  it should "get empty list when no linked accounts" in withMinimalTestDatabaseAndServices { service =>
+    when(service.bondApiDAO.getBondProviders()).thenReturn(Future.successful(List("p1", "p2")))
+    when(service.bondApiDAO.getServiceAccountKey("p1", userInfo)).thenReturn(Future.successful(None))
+    when(service.bondApiDAO.getServiceAccountKey("p2", userInfo)).thenReturn(Future.successful(None))
+
+    service.getBondProviderServiceAccountEmails(userInfo).futureValue shouldBe List.empty
+  }
+
+  "grantRequesterPaysToLinkedSAs" should "link" in withMinimalTestDatabaseAndServices { service =>
+    val expectedEmail = BondServiceAccountEmail("bondSA")
+
+    when(service.bondApiDAO.getBondProviders()).thenReturn(Future.successful(List("p1", "p2")))
+    when(service.bondApiDAO.getServiceAccountKey("p1", userInfo)).thenReturn(Future.successful(None))
+    when(service.bondApiDAO.getServiceAccountKey("p2", userInfo)).thenReturn(Future.successful(Some(BondResponseData(expectedEmail))))
+
+    runAndWait(workspaceRequesterPaysQuery.userExistsInWorkspaceNamespace(minimalTestData.billingProject.projectName.value, userInfo.userEmail)) shouldBe false
+    service.grantRequesterPaysToLinkedSAs(userInfo, minimalTestData.workspace.toWorkspaceName).futureValue shouldBe List(expectedEmail)
+    runAndWait(workspaceRequesterPaysQuery.userExistsInWorkspaceNamespace(minimalTestData.billingProject.projectName.value, userInfo.userEmail)) shouldBe true
+    service.googleServicesDAO.asInstanceOf[MockGoogleServicesDAO].policies.get(minimalTestData.billingProject.projectName) shouldBe Some(Map(service.requesterPaysRole -> Set("serviceAccount:" + expectedEmail.client_email)))
+
+    // second call should not fail
+    service.grantRequesterPaysToLinkedSAs(userInfo, minimalTestData.workspace.toWorkspaceName).futureValue shouldBe List(expectedEmail)
+  }
+
+  "revokeUserFromWorkspace" should "unlink" in withMinimalTestDatabaseAndServices { service =>
+    val expectedEmail = BondServiceAccountEmail("bondSA")
+
+    // add user to 2 workspaces in same namespace
+    runAndWait(workspaceRequesterPaysQuery.insertAllForUser(minimalTestData.workspace.toWorkspaceName, userInfo.userEmail, Set(expectedEmail)))
+    runAndWait(workspaceRequesterPaysQuery.insertAllForUser(minimalTestData.workspace2.toWorkspaceName, userInfo.userEmail, Set(expectedEmail)))
+    runAndWait(workspaceRequesterPaysQuery.userExistsInWorkspaceNamespace(minimalTestData.billingProject.projectName.value, userInfo.userEmail)) shouldBe true
+
+    // add user to mock google bindings
+    val initialBindings = Map(service.requesterPaysRole -> Set("serviceAccount:" + expectedEmail.client_email))
+    service.googleServicesDAO.asInstanceOf[MockGoogleServicesDAO].policies.put(minimalTestData.billingProject.projectName, initialBindings)
```

Comment:
Why do we need to manually add the user to the google policy bindings when adding the user to the workspace, but this happens automatically when deleting the user from the workspace?

Reply:
It is more like just in time triggering for addition. The trigger will be starting a workflow or notebook. Doing it when adding to a workspace is not great because the user may not have linked accounts yet (or even be a registered user for that matter).